90.md

```diff
@@ -0,0 +1,171 @@
+
+This NIP is not to be confused with a 1:1 marketplace; instead, it describes a flow where a user announces a desired output, willingness to pay, and service providers compete to fulfill the job requirement in the best way possible.
+
+### Actors
+There are two actors in the workflow described in this NIP:
+* Customers (npubs who request a job)
+* Service providers (npubs who fulfill jobs)
+
+## Job request (`kind:5000-5999`)
+A request to process data, published by a customer. This event signals that an customer is interested in receiving the result of some kind of compute.
+
+```json
+{
+    "kind": 5xxx, // kind in 5000-5999 range
+    "content": "",
+    "tags": [
+        [ "i", "<data>", "<input-type>", "<relay>", "<marker>" ],
+        [ "output", "<mime-type>" ],
+        [ "relays", "wss://..."],
+        [ "bid", "<msat-amount>" ],
+        [ "t", "bitcoin" ]
+    ]
+}
+```
+
+All tags are optional.
+
+* `i` tag: Input data for the job (zero or more inputs)
+    * `<data>`: The argument for the input
+    * `<input-type>`: The way this argument should be interpreted. MUST be one of:
+        * `url`: A URL to be fetched of the data that should be processed.
+        * `event`: A Nostr event ID.
+        * `job`: The output of a previous job with the specified event ID. The dermination of which output to build upon is up to the service provider to decide (e.g. waiting for a signaling from the customer, waiting for a payment, etc.)
+        * `text`: `<data>` is the value of the input, no resolution is needed
+    * `<relay>`: If `event` or `job` input-type, the relay where the event/job was published, otherwise optional or empty string
+    * `<marker>`: An optional field indicating how this input should be used within the context of the job
+* `output`: Expected output format. Different job request `kind` defines this more precisely.
+* `param`: Optional parameters for the job as key (first argument)/value (second argument). Different job request `kind` defines this more precisely. (e.g. `[ "param", "lang", "es" ]`)
+* `bid`: Customer MAY specify a maximum amount (in millisats) they are willing to pay
+* `relays`: List of relays where Service Providers SHOULD publish responses to
+* `p`: Service Providers the customer is interested in. Other SPs MIGHT still choose to process the job
+
+## Job result (`kind:6000-6999`)
+
+Service providers publish job results, providing the output of the job result. They should tag the original job request event id as well as the customer's pubkey.
+
+```json
+{
+    "pubkey": "<service-provider pubkey>",
+    "content": "<payload>",
+    "kind": 6xxx,
+    "tags": [
+        [ "request", "<job-request>" ],
+        [ "e", "<job-request-id>", "<relay-hint>" ],
+        [ "i", "<input-data>" ],
```

Comment:
What is the rationale of introducing input tags to the job result if they could be derived from the request tag? I don't dislike it, just wondering if output tag (or any other tags really) should get the same treatment too?

Reply:
the idea is that any app can very quickly see if there is a job result for the same input it would have provided, this might seem trivial, but for things where speed matters for UX, saving a round trip is important.

The example I have in mind is translation, which is the case where I see this having the largest immediate impact; instead of pinging a translation server one time per user for the same note ID, clients could query for translations of the event ids they want to see a translation for in a single REQ.

`REQ, .., { kinds: [<translation-result-kind>], "#i": [ "event_ids"... ] }`